### ISSUE TITLE
about narenas

### ISSUE THREAD

**Issue by wangliguang111** [jemalloc/jemalloc#2758]
"I'm using jemalloc, and I've set the following environment variables:

ENV LD_PRELOAD=`jemalloc-config --libdir`/usr/lib/libjemalloc.so`jemalloc-config --revision`
ENV MALLOC_CONF="narenas:20"
I have a few questions:

Does the configuration option narenas refer to the maximum number of arenas that the operating system can partition? The official explanation, 'Maximum number of arenas to use for automatic multiplexing of threads and arenas,' is somewhat vague.

I've noticed that in performance testing, the parameter MALLOC_ARENA_MAX still seems to have an effect, as the performance differs from when only narenas is set. Could you clarify this?"

**Comment by guangli-dai** [comment]
The arena here is an internal level of jemalloc, it has nothing to do with the operating system partitions. Each arena can individually serve requests (malloc/free) and one arena is expected to serve multiple threads. Adding the number of arenas can help reduce CPU usage by reducing contentions while adding the overhead in memory usage.

> the parameter MALLOC_ARENA_MAX still seems to have an effect

As far as I know, `MALLOC_ARENA_MAX` is a glibc parameter and tunes the behavior of libc malloc instead of jemalloc. If it somehow still plays a role with jemalloc linked, maybe jemalloc only replace it partially.

**Comment by wangliguang111** [comment]
Thank you for your answer. What is the maximum number of arenas that jemalloc can utilize (not just for a single thread)? Is it determined by narenas?

**Comment by guangli-dai** [comment]
> Thank you for your answer. What is the maximum number of arenas that jemalloc can utilize (not just for a single thread)? Is it determined by narenas?

Yes, `narenas` is the maximum number of auto arenas. If you haven't created any manual ones, the only exception would be the oversize arena, which is only one more (it may or may not be there since I don't know which version you are on). If you are on the latest version, [`narenas_ratio`](https://github.com/jemalloc/jemalloc/blob/b82333fdec6e5833f88780fcf1fc50b799268e1b/src/jemalloc.c#L1311) may override `narenas` when you set it, so just set either one to prevent confusion.
